src/print.c

```diff
@@ -170,43 +173,111 @@ print_dialog_handle_close (PrintDialogHandle *handle)
   print_dialog_handle_free (handle);
 }
 
+static gboolean
+can_preview (void)
+{
+  g_autofree char *path = NULL;
+
+  path = g_find_program_in_path ("evince-previewer");
+
+  if (path)
+    return TRUE;
+
+  g_warning ("evince-previewer not found, disabling print preview");
+
+  return FALSE;
+}
+
+static gboolean
+launch_preview (const char *filename,
+                const char *title,
+                GtkPrintSettings *settings,
+                GtkPageSetup *page_setup,
+                GError **error)
+{
+  g_autofree char *cmd = NULL;
+  g_autoptr(GAppInfo) appinfo = NULL;
+  g_autoptr(GAppLaunchContext) context = NULL;
+  gboolean retval = FALSE;
+  g_autoptr(GKeyFile) keyfile = NULL;
+  g_autofree char *data = NULL;
+  gsize data_len;
+  g_autofree char *settings_filename = NULL;
+  int fd;
+
+  fd = g_file_open_tmp ("settingsXXXXXX.ini", &settings_filename, error);
+
+  keyfile = g_key_file_new ();
+  if (settings)
+    gtk_print_settings_to_key_file (settings, keyfile, NULL);
+  if (page_setup)
+    gtk_page_setup_to_key_file (page_setup, keyfile, NULL);
+
+  g_key_file_set_string (keyfile, "Print Job", "title", title);
+
+  data = g_key_file_to_data (keyfile, &data_len, NULL);
+  if (data)
+    g_file_set_contents (settings_filename, data, data_len, NULL);
+
+  cmd = g_strdup_printf ("evince-previewer --unlink-tempfile --print-settings %s %s", settings_filename, filename);
+
+  g_debug ("launching %s", cmd);
+
+  appinfo = g_app_info_create_from_commandline (cmd,
+                                                "Print Preview",
+                                                G_APP_INFO_CREATE_NONE,
+                                                NULL);
+  if (!appinfo)
+    goto out;
+
+  context = (GAppLaunchContext *)gdk_display_get_app_launch_context (gdk_display_get_default ());
+  retval = g_app_info_launch (appinfo, NULL, G_APP_LAUNCH_CONTEXT (context), error);
+
+out:
+  if (fd > 0)
+    close (fd);
+
+  return retval;
+}
+
 static gboolean
 print_file (int fd,
             const char *app_id,
+            gboolean preview,
             GtkPrinter *printer,
             GtkPrintSettings *settings,
             GtkPageSetup *page_setup,
             GError **error)
 {
-  GtkPrintJob *job;
+  GtkPrintJob *job = NULL;
```

Comment:
There is an autoptr definition for GtkPrintJob in https://gitlab.gnome.org/GNOME/gtk/blob/master/gtk/gtkunixprint-autocleanups.h#L24, so I think you can use that here (and get rid of those unrefs)